File: macos/ValveProvisioningApp/Sources/ValveDomain/Validation.swift
import Foundation

public enum ValidationError: Error, LocalizedError, Sendable
{ case missingField(String)
  case invalidPlatform
  case invalidPublicKey

  public var errorDescription: String?
  { let description: String
    switch self
    { case .missingField(let field): description = "\(field) is required."
      case .invalidPlatform: description = "Platform must be macOS."
      case .invalidPublicKey: description = "Public key must be base64 with 32 bytes."
    }
    return description
  }
}

public enum CredentialRequestValidator
{ public static func validateRegister(_ request: RegisterCredentialRequest) throws
  { try validateCommonFields(tenantID: request.tenantID, actorUserID: request.actorUserID, installID: request.installID)
    if request.platform != "macOS" { throw ValidationError.invalidPlatform }
    if request.appBundleID.isEmpty { throw ValidationError.missingField("app_bundle_id") }
    if request.appVersion.isEmpty { throw ValidationError.missingField("app_version") }
    if request.appBuild.isEmpty { throw ValidationError.missingField("app_build") }
    if request.deviceLabel.isEmpty { throw ValidationError.missingField("device_label") }
    if !isValidPublicKeyBase64(request.publicKey) { throw ValidationError.invalidPublicKey }
  }

  public static func validateRotate(_ request: RotateCredentialRequest) throws
  { try validateCommonFields(tenantID: request.tenantID, actorUserID: request.actorUserID, installID: request.installID)
    if request.oldCredentialID.isEmpty { throw ValidationError.missingField("old_credential_id") }
    if request.appVersion.isEmpty { throw ValidationError.missingField("app_version") }
    if request.appBuild.isEmpty { throw ValidationError.missingField("app_build") }
    if request.deviceLabel.isEmpty { throw ValidationError.missingField("device_label") }
    if !isValidPublicKeyBase64(request.newPublicKey) { throw ValidationError.invalidPublicKey }
  }

  public static func validateRevoke(_ request: RevokeCredentialRequest) throws
  { try validateCommonFields(tenantID: request.tenantID, actorUserID: request.actorUserID, installID: nil)
    if request.credentialID.isEmpty { throw ValidationError.missingField("credential_id") }
    if request.reason.isEmpty { throw ValidationError.missingField("reason") }
  }

  private static func validateCommonFields(tenantID: String, actorUserID: String, installID: String?) throws
  { if tenantID.isEmpty { throw ValidationError.missingField("tenant_id") }
    if actorUserID.isEmpty { throw ValidationError.missingField("actor_user_id") }
    if let installID, installID.isEmpty { throw ValidationError.missingField("install_id") }
  }

  public static func isValidPublicKeyBase64(_ value: String) -> Bool
  { let data = Data(base64Encoded: value)
    let isValid = data?.count == 32
    return isValid
  }
}

File: internal/credentials/validation.go
package credentials

import (
	"crypto/ed25519"
	"encoding/base64"
	"errors"
	"fmt"
)

// #R001: Validate register payloads, mode constraints, and key requirements.
func ValidateRegister(req RegisterRequest, allowEmptyActor bool, hmacEnabled bool) error {
	if req.TenantID == "" {
		return errors.New("tenant_id is required")
	}
	if req.ActorUserID == "" && !allowEmptyActor {
		return errors.New("actor_user_id is required")
	}
	if req.InstallID == "" {
		return errors.New("install_id is required")
	}
	if req.AppBundleID == "" {
		return errors.New("app_bundle_id is required")
	}
	if req.Platform != "macOS" {
		return errors.New("platform must be macOS")
	}
	if req.CredentialMode != ModeEd25519 && req.CredentialMode != ModeHMACSHA256 {
		return errors.New("credential_mode must be ed25519 or hmac_sha256")
	}
	if req.CredentialMode == ModeHMACSHA256 && !hmacEnabled {
		return errors.New("hmac mode is disabled")
	}
	if req.CredentialMode == ModeEd25519 {
		if req.PublicKey == "" {
			return errors.New("public_key is required for ed25519 mode")
		}
		decoded, err := base64.StdEncoding.DecodeString(req.PublicKey)
		if err != nil {
			return fmt.Errorf("public_key must be base64 encoded: %w", err)
		}
		if len(decoded) != ed25519.PublicKeySize {
			return errors.New("public_key has invalid ed25519 key length")
		}
	}
	return nil
}

// #R005: Validate revoke payload identifiers and actor policy constraints.
func ValidateRevoke(req RevokeRequest, allowEmptyActor bool) error {
	if req.TenantID == "" {
		return errors.New("tenant_id is required")
	}
	if req.ActorUserID == "" && !allowEmptyActor {
		return errors.New("actor_user_id is required")
	}
	if req.CredentialID == "" {
		return errors.New("credential_id is required")
	}
	return nil
}

// #R010: Validate rotation payloads including mode-specific replacement key checks.
func ValidateRotate(req RotateRequest, allowEmptyActor bool, hmacEnabled bool) error {
	if req.TenantID == "" {
		return errors.New("tenant_id is required")
	}
	if req.ActorUserID == "" && !allowEmptyActor {
		return errors.New("actor_user_id is required")
	}
	if req.OldCredentialID == "" {
		return errors.New("old_credential_id is required")
	}
	if req.InstallID == "" {
		return errors.New("install_id is required")
	}
	if req.CredentialMode != ModeEd25519 && req.CredentialMode != ModeHMACSHA256 {
		return errors.New("credential_mode must be ed25519 or hmac_sha256")
	}
	if req.CredentialMode == ModeHMACSHA256 && !hmacEnabled {
		return errors.New("hmac mode is disabled")
	}
	if req.CredentialMode == ModeEd25519 {
		if req.NewPublicKey == "" {
			return errors.New("new_public_key is required for ed25519 mode")
		}
		decoded, err := base64.StdEncoding.DecodeString(req.NewPublicKey)
		if err != nil {
			return fmt.Errorf("new_public_key must be base64 encoded: %w", err)
		}
		if len(decoded) != ed25519.PublicKeySize {
			return errors.New("new_public_key has invalid ed25519 key length")
		}
	}
	return nil
}

// #R015: Validate upload target request identifiers before service dispatch.
func ValidateUploadTargetRequest(req UploadTargetRequest) error {
	if req.TenantID == "" {
		return errors.New("tenant_id is required")
	}
	if req.InstallID == "" {
		return errors.New("install_id is required")
	}
	if req.CredentialID == "" {
		return errors.New("credential_id is required")
	}
	return nil
}
